Add Gateway API conformance tests to CI pipeline

electricjesus opened the pull request:
## Description

This PR adds Gateway API conformance testing to the Calico CI pipeline.

Automated conformance testing to ensure Calico's Gateway API implementation meets the standard specification and to catch regressions early.

Installs latest calicoreates an `operator.tigera.io/GatewayAPI` resource with matching `gateway.networking.k8s.io/Gateway` then Gateway Conformance tests are run.

## Related issues/PRs

N/A

## Todos

- [x] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

caseydavenport commented on .semaphore/semaphore.yml.d/blocks/20-api.yml:
```diff
@@ -1,15 +0,0 @@
-- name: API
```
@electricjesus was it intentional to remove all of these blocks? Seems wrong?

electricjesus committed:
deps: pin sigs.k8s.io/gateway-api v1.4.1

Adds the upstream Gateway API conformance suite as a direct dependency
so we can run it against Calico's Envoy-Gateway-based implementation
and emit ConformanceReport YAMLs for upstream submission.

v1.4.1 matches the Gateway API version compiled into Envoy Gateway
v1.7.x (per https://gateway.envoyproxy.io/news/releases/matrix/),
which is the version pinned by third_party/envoy-gateway/Makefile on
master and on release-v3.31. Submissions for release-v3.30 (EG v1.5.9
-> Gateway API v1.3.0) will need a separate dep bump on that branch.

electricjesus committed:
e2e/gateway: add Gateway API conformance runner

Adds a thin test wrapper around sigs.k8s.io/gateway-api/conformance
that runs the upstream suite against the cluster at \$KUBECONFIG and
writes a ConformanceReport YAML to -report-output, suitable for
submission to kubernetes-sigs/gateway-api/conformance/reports/.

The runner reuses upstream's flag definitions
(sigs.k8s.io/gateway-api/conformance/utils/flags) so every knob the
suite supports is surfaced without local duplication: -gateway-class,
-conformance-profiles, -mode, -supported-features, -exempt-features,
-all-features, -skip-tests, -run-test, -allow-crds-mismatch, plus
the ConformanceReport metadata flags (-organization, -project, -url,
-contact, -version).

Wires the embedded conformance.Manifests fs.FS so the suite's base
manifests get applied without extra setup, and registers the
apiextensions/v1 scheme so the suite's CRD-version sanity check
(used by -allow-crds-mismatch) can list CustomResourceDefinitions
through the controller-runtime client.

Ships a minimal operator.tigera.io/v1 GatewayAPI custom resource at
manifests/gatewayapi.yaml. When applied, the tigera-operator
provisions the default GatewayClass "tigera-gateway-class" in the
"tigera-gateway" namespace, which is what the conformance run targets
via -gateway-class.

electricjesus committed:
build: wire Gateway API conformance into Make targets

Adds three new top-level targets that mirror the existing e2e
patterns (e2e-test, e2e-run-cnp) and produce a ConformanceReport YAML
for upstream submission.

  e2e-gateway-setup          - apply manifests/gatewayapi.yaml and
                               wait for the default GatewayClass to
                               reach Accepted=true.

  e2e-run-gateway-conformance - run e2e/bin/gateway/e2e.test against
                                \$KUBECONFIG with full upstream flag
                                passthrough; emits the report YAML to
                                \$GATEWAY_CONFORMANCE_REPORT.

  e2e-test-gateway-conformance - one-shot: build the binary, kind-up,
                                 setup, run.

All knobs are tunable as Make variables for multi-version submissions
(release-v3.30 ships Gateway API v1.3.0; release-v3.31 / release-v3.32
ship v1.4.1):

  GATEWAY_CONFORMANCE_VERSION   - Calico release tag (e.g. v3.31.5).
                                  REQUIRED for upstream submission;
                                  must be a tag/SHA per
                                  conformance/reports/README.md.
  GATEWAY_CLASS_NAME            - default tigera-gateway-class.
  GATEWAY_CONFORMANCE_PROFILES  - default GATEWAY-HTTP.
  GATEWAY_CONFORMANCE_MODE      - default default.
  GATEWAY_CONFORMANCE_REPORT    - report YAML output path.
  GATEWAY_CONFORMANCE_{ORG,PROJECT,URL,CONTACT}
                                - ConformanceReport.implementation
                                  metadata.

Adds bin/gateway/e2e.test to e2e/Makefile's build target. Unlike
clusternetworkpolicy, this binary needs no CGO/libbpf - it speaks to
the cluster via client-go and runs the upstream suite, so the build
recipe uses DOCKER_RUN (not DOCKER_RUN_CGO).

electricjesus committed:
ci: add Semaphore block for Gateway API conformance tests

Adds a path-scoped CI block that runs the upstream Gateway API
conformance suite against a Calico-deployed kind cluster on every PR
that touches the conformance runner or any of the third_party/envoy-*
sources. 30-minute time limit; depends on the existing Prerequisites
block.

The block builds e2e/bin/gateway/e2e.test, brings up the kind cluster
via the canonical `make kind-up` flow (which deploys Calico via Helm
through the tigera-operator), applies the operator GatewayAPI CR via
`make e2e-gateway-setup`, then runs `make e2e-run-gateway-conformance`
to execute the upstream suite and emit the ConformanceReport YAML.
GATEWAY_CONFORMANCE_VERSION is sourced from SEMAPHORE_GIT_TAG_NAME
(release tag) or SEMAPHORE_GIT_BRANCH so the report's
implementation.version field reflects the build under test.

The epilogue publishes JUnit XML and pushes the ./report/ directory as
a per-job Semaphore artifact named gateway-conformance-reports for
later upstream submission.

Trigger paths intentionally omit the operator CRDs / charts since
those are upstream-defined and the tigera-operator pulls them through
its own release flow; only changes to the runner, manifests, or the
bundled envoy-* sources rebuild and rerun this suite.

Note: regeneration of .semaphore/semaphore.yml and
.semaphore/semaphore-scheduled-builds.yml from this template happens
in a follow-up commit so the diff stays reviewable.

electricjesus committed:
generate: regenerate semaphore yamls and deps.txt files

Picks up the new gateway conformance Semaphore block in
.semaphore/semaphore.yml.d/blocks/20-gateway-conformance.yml and
the sigs.k8s.io/gateway-api dependency added by the conformance
runner. Output of:

  make gen-semaphore-yaml
  make gen-deps-files

Per repo convention these generated files travel alongside their
source changes.

electricjesus committed:
build: include envoy images in kind-build-images flow

Today the third_party/envoy-{gateway,proxy,ratelimit} Makefiles build
local images named envoy-<component>:latest-\$(ARCH), with no
calico/-prefixed retag, and lib.Makefile's kind-build-images flow
doesn't trigger them. As a result, when the operator-driven kind
deploy resolves the calico_versions.yml entries (envoy-gateway:
test-build, envoy-proxy: test-build, envoy-ratelimit: test-build)
into image references calico/envoy-<component>:test-build, those
images don't exist locally and the kind nodes fail to pull them
(PullNever in test mode).

Wire the three envoy components into the same build/load loop as the
core Calico images so a single `make kind-up` produces a fully
functional cluster with the bundled envoy versions installed:

  - third_party/envoy-{gateway,proxy,ratelimit}/Makefile: change the
    default image name to calico/envoy-<component> so the build
    target produces calico/envoy-<component>:latest-\$(ARCH), matching
    the pattern node/whisker/cmd/calico use.

  - lib.Makefile: extend KIND_CALICO_IMAGES to include the three new
    test-build references and KIND_IMAGE_MARKERS to depend on each
    third_party/envoy-<component>/.envoy-<component>.created-\$(ARCH)
    stamp. The existing retag loop in kind-build-images then re-tags
    each :latest-\$(ARCH) image as :test-build automatically.

Side benefit: the Gateway API conformance CI block already triggers
on changes under /third_party/envoy-*, so a version bump in any of
those Makefiles now automatically rebuilds the bundled image, deploys
it onto the kind cluster, and re-runs the conformance suite. One less
moving piece to remember when bumping the bundled Envoy Gateway.

electricjesus committed:
fix(e2e-gateway-setup): poll for Gateway API CRDs before wait

The tigera-operator reconciles the operator.tigera.io/v1 GatewayAPI CR
asynchronously: it installs the Gateway API CRDs (gatewayclasses,
gateways, httproutes, ...), installs the Envoy Gateway controller,
and only then creates the default GatewayClass. The previous recipe
called `kubectl wait --for=condition=Accepted gatewayclass/...`
immediately after applying the CR, which fails fast (not retries)
when the resource type itself doesn't yet exist server-side, so a CI
job that beat the operator to the wait would always error with:

  error: the server doesn't have a resource type "gatewayclass"

Poll for the gatewayclasses.gateway.networking.k8s.io CRD first
(default 5m), then poll for the GatewayClass object to be created
(default 5m), then wait on its Accepted condition with the existing
timeout. Each phase emits diagnostics on timeout (operator logs,
tigerastatus, full GatewayAPI CR yaml) so a CI failure has enough
context to triage from logs alone.

GATEWAY_SETUP_CRD_TIMEOUT (seconds) and GATEWAY_SETUP_GWC_TIMEOUT
(kubectl wait duration) are tunable for local dev where the operator
can take longer or shorter to reconcile.

electricjesus committed:
fix(e2e-run-gateway-conformance): default to all upstream features

The tigera-operator-provisioned GatewayClass tigera-gateway-class
doesn't populate .status.supportedFeatures (a v1.4+ Gateway API
field that implementations are expected to advertise). The upstream
conformance suite's auto-inference reads that field, finds the empty
set, and refuses to construct with:

  Supported features for GatewayClass tigera-gateway-class: []
  constructing conformance suite: no supported features were
  determined for test suite

Pass -all-features=true to the runner so it uses the full upstream
features.AllFeatures set instead of inferring from the GatewayClass
status. Profile core-features get unioned in by the suite as before.

Add three new tunables:

  GATEWAY_CONFORMANCE_ALL_FEATURES (default true)  - flips the
    runner between "use AllFeatures" and "use whatever -supported-
    features says (or infer from GWC status)".

  GATEWAY_CONFORMANCE_SUPPORTED_FEATURES (default empty) - explicit
    comma-separated feature list. Submission-quality reports should
    set ALL_FEATURES=false and curate this to match what Calico-
    deployed Envoy Gateway actually supports.

  GATEWAY_CONFORMANCE_EXEMPT_FEATURES (default empty) - features to
    subtract from the supported set; useful for excluding mesh
    profiles which the GATEWAY-HTTP profile shouldn't exercise.

For the CI smoke test, all-features=true is the cheapest way to get
a pass/fail signal across the whole suite. Real upstream submissions
should switch to a curated list (see Envoy Gateway's
test/conformance/suite.go for the canonical curated set, since
Calico's implementation is just a Calico-deployed Envoy Gateway).

electricjesus committed:
e2e/gateway: add curated feature/test/profile sets, default to envoy-gateway

Calico's Gateway API implementation is a Calico-deployed Envoy
Gateway, so its supported feature surface is identical to upstream
envoyproxy/gateway's. Rather than maintaining a parallel curated list
in this repo, mirror EG's official conformance suite configuration
(envoyproxy/gateway/internal/gatewayapi/conformance/suite.go at the
v1.7.0 tag) inside the runner and select it via -curated=envoy-gateway.

The curated set comprises:

  SupportedFeatures = features.AllFeatures
                    + features.UDPRouteFeatures
                    - {GatewayStaticAddresses,
                       GatewayInfrastructurePropagation}

  ExemptFeatures    = features.MeshCoreFeatures
                    + features.MeshExtendedFeatures
                    (Envoy Gateway has no service-mesh support, so
                    Calico inherits that gap)

  SkipTests         = [GatewayStaticAddresses, GatewayInfrastructure]

  Profiles          = {GATEWAY-HTTP, GATEWAY-TLS, GATEWAY-GRPC}

When -curated is non-empty the four upstream flag-driven knobs
(-supported-features, -exempt-features, -skip-tests, -all-features,
-conformance-profiles) are overridden so the caller can't accidentally
mix and match an inconsistent set. -curated="" preserves the previous
ad-hoc behaviour for debugging runs.

Default the make target to GATEWAY_CONFORMANCE_CURATED=envoy-gateway
so CI gets honest per-profile numbers without curating-by-hand. Add
GATEWAY_CONFORMANCE_CURATED as a tunable so a future release-tag
submission can pin against a different EG version's curated set if
needed (or override to empty for ad-hoc testing).

electricjesus committed:
fix(e2e-gateway-setup): add L2 metallb pool for host-reachable LB IPs

The Gateway API conformance test binary runs on the host (Semaphore VM
or developer laptop) and makes raw HTTP requests to the Gateway's
external LoadBalancer IP. Calico's default metallb pool
(hack/test/kind/infra/metallb-config.yaml) advertises 80.15.0.0/24 in
BGP mode, which Calico's in-cluster BIRD fans out as cluster routes.
Pods inside the cluster can reach those IPs; the host cannot, because
it has no BGP session with metallb. Result: every conformance HTTP
request times out:

  Request failed, not ready yet: Get "http://80.15.0.4/...":
  context deadline exceeded

Add a SECOND metallb pool in L2 mode whose addresses come from the
kind docker bridge subnet. Hosts have a direct route to that bridge,
so LB IPs land on the bridge and are reachable from the test runner
without BGP. The pool is namespace-scoped to the conformance test
namespaces (gateway-conformance-{infra,app-backend,web-backend} plus
tigera-gateway in case the operator places EG-managed Services in the
controller's namespace) and has serviceAllocation.priority=100 so
metallb deterministically chooses this pool over the default BGP one
for those Services. Other tests' LoadBalancer Services in unrelated
namespaces continue to use the BGP pool unchanged.

This mirrors what envoyproxy/gateway's tools/hack/create-cluster.sh
does for its own conformance CI: query `docker network inspect kind`
for the bridge IPv4 subnet, take .200-.250 of the same /24, advertise
in L2 mode. The addresses are computed at apply time inside the
e2e-gateway-setup target so the manifest stays a static template with
__LB_RANGE_V4__ as a placeholder.

electricjesus committed:
build: upgrade kind metallb to v0.14.9 (CRD-based config)

Calico's kind cluster bringup shipped metallb in legacy ConfigMap-mode
(`address-pools` under a single `config` ConfigMap, applied via
`metallb.yaml` with no CRDs). metallb deprecated that mode in v0.13
and removed support for new features there. Two concrete reasons to
modernize now:

1. The Gateway API conformance test runner (e2e-gateway-setup, added
   in 9818ff403d) needs a SECOND, namespace-scoped pool in L2 mode
   alongside the existing default BGP pool. Per-namespace
   serviceAllocation only exists in CRD-based config; legacy
   ConfigMap mode has no equivalent. The L2 pool YAML uses
   metallb.io/v1beta1 IPAddressPool + L2Advertisement, so the
   target was failing with `no matches for kind "IPAddressPool"`.

2. metallb-native.yaml (the CRD-based controller manifest) is what
   the upstream Envoy Gateway project itself uses for its own kind-
   based conformance CI. Aligning means a future maintainer can
   compare like-for-like with EG's bringup when debugging.

Changes:

  - hack/test/kind/infra/metallb.yaml: vendor metallb v0.14.9's
    metallb-native.yaml (CRDs + controller Deployment + speaker
    DaemonSet + ValidatingWebhookConfiguration).
    Source: https://raw.githubusercontent.com/metallb/metallb/v0.14.9/config/manifests/metallb-native.yaml

  - hack/test/kind/infra/metallb-config.yaml: convert the existing
    ConfigMap-driven `default` BGP pool to a metallb.io/v1beta1
    IPAddressPool + BGPAdvertisement pair. Same address ranges
    (80.15.0.0/24 and fdff::/64), same BGP semantics — preserves all
    existing test behaviour.

  - hack/test/kind/deploy_resources.sh: after applying metallb.yaml,
    wait for the controller Deployment to be ready and for the
    IPAddressPool / BGPAdvertisement / L2Advertisement CRDs to be
    Established. Then retry the metallb-config.yaml apply for ~60s,
    because the ValidatingWebhookConfiguration takes a couple seconds
    to start serving even after the Deployment reports Ready (metallb
    rejects kubectl-apply during that window with
    "failed calling webhook ... no endpoints available").

Tests that were depending on BGP-advertised LB IPs being reachable
from in-cluster pods continue to work unchanged (same pool name,
same addresses, same protocol). The only consumer-visible change is
that `kubectl edit cm config -n metallb-system` to retune the pool
during dev no longer works — `kubectl edit ipaddresspool default -n
metallb-system` is the new equivalent.

electricjesus committed:
fix(e2e-gateway-setup): override Envoy data-plane Service to externalTrafficPolicy=Cluster

Envoy Gateway upstream defaults the Gateway data-plane Service's
externalTrafficPolicy to "Local" (api/v1alpha1/shared_types.go in
envoyproxy/gateway, +kubebuilder:default:="Local"). With L2-mode
metallb only one speaker node claims any given LB IP at a time;
metallb's L2 selection is Local-aware, but in a multi-node kind
cluster (1 control + 3 workers, the Calico kind default) some tests
still land on a speaker node that doesn't host the matching backend
pod, and IPVS then drops the packet because the per-node Service has
zero local endpoints. Concretely on this PR's GCP-VM repro,
externalTrafficPolicy=Local left these tests failing in the
non-deterministic-routing band:

  BackendTLSPolicyConflictResolution
  BackendTLSPolicyInvalidCACertificateRef
  BackendTLSPolicyInvalidKind
  HTTPRouteCrossNamespace

Switching the data-plane Service to externalTrafficPolicy=Cluster
clears all four (66 PASS / 8 FAIL became 70 PASS / 4 FAIL on the same
suite invocation; the remaining failures are unrelated TLSRoute /
UDPRoute / ReferenceGrant tests we'll pursue separately).

Envoy Gateway's own conformance CI gets away with the Local default
because tools/hack/create-cluster.sh in envoyproxy/gateway defaults
NUM_WORKERS to 0 — single-node kind, where the speaker and the
backend pod always share a node.

The override flows through Calico's tigera-operator GatewayAPI CR via
envoyProxyRef → a base EnvoyProxy that customises envoyService. The
operator copies this into the auto-generated per-GatewayClass
EnvoyProxy on each reconcile, so the override is durable across
reconciliations rather than racing with the operator like a kubectl
patch would.

Files:

  - e2e/cmd/gateway/manifests/envoyproxy.yaml: new EnvoyProxy
    "gateway-conformance-base" in tigera-gateway with
    spec.provider.kubernetes.envoyService.externalTrafficPolicy=Cluster.

  - e2e/cmd/gateway/manifests/gatewayapi.yaml: add envoyProxyRef on
    the tigera-gateway-class GatewayClass entry pointing at the new
    EnvoyProxy.

  - Makefile: e2e-gateway-setup applies the GatewayAPI CR first
    (which triggers the operator to create the tigera-gateway
    namespace and install the EnvoyProxy CRD), waits for the
    namespace + CRD, then applies the EnvoyProxy. The operator's
    reconcile loop requeues with a missing-EnvoyProxy error in the
    interim, which resolves on the next pass.

CI execution time may grow modestly because of the extra requeue
window before the GatewayClass becomes Accepted; we'll measure on
this PR's run and tune GATEWAY_SETUP_GWC_TIMEOUT if needed.

electricjesus committed:
fix(e2e-gateway-setup): pre-create tigera-gateway namespace to break cold-start deadlock

CI's first reconcile hung at "EnvoyProxy ... not found" and the wait
loop timed out waiting for the tigera-gateway namespace.

The gatewayapi controller in tigera-operator renders CRDs early
(gatewayapi_controller.go:221) but renders the tigera-gateway namespace
plus all non-CRD resources only AFTER resolving every envoyProxyRef
(line 419, behind the Get at line 345-351). When envoyProxyRef is set on
first install the EnvoyProxy CR doesn't exist yet, so SetDegraded fires
at line 350 and Reconcile returns early. CRDs are applied; namespace
isn't. The next reconcile loops the same way, and there's no path that
ever creates the namespace, which means we can never apply the
EnvoyProxy CR there, which means the deadlock never breaks.

The previous setup target waited for the namespace to appear and so
hung for 300s before failing. The VM-side dev loop didn't hit this
because its cluster was reconciled in stages and the namespace was
already there before envoyProxyRef was added.

Wait only on the EnvoyProxy CRD being established, then create the
namespace ourselves and apply the EnvoyProxy CR. The next reconcile
finds the EnvoyProxy and proceeds normally; the operator's non-CRD
render takes ownership of the existing namespace via standard
CreateOrUpdateOrDelete semantics.

Manifest comments updated to describe the new bootstrap order.

electricjesus committed:
fix(e2e/gateway): register all gateway-api API versions in client scheme

The four remaining conformance failures (HTTPRouteReferenceGrant,
TLSRouteInvalidReferenceGrant, TLSRouteSimpleSameNamespace, UDPRoute)
share one root cause. The runner's controller-runtime client only knew
about gateway-api v1, so any test that touched a type defined in another
version failed with:

  no kind is registered for the type v1beta1.ReferenceGrant in scheme
  "pkg/runtime/scheme.go:111"

ReferenceGrant lives in v1beta1; TLSRoute and UDPRoute in v1alpha2;
BackendTLSPolicy in v1alpha3. HTTPRouteReferenceGrant deletes the
ReferenceGrant mid-test, the TLS tests apply manifests with TLSRoute,
the UDP test applies UDPRoute — every one of them ended up calling into
the typed client for a kind it couldn't decode.

Mirror sigs.k8s.io/gateway-api/conformance.DefaultOptions and install
v1alpha3, v1alpha2, v1beta1, and v1 alongside apiextensions/v1.

Verified focused on a kind cluster: all four previously-failing tests
now pass.

electricjesus committed:
ci(gateway-conformance): publish report to GCS for permanent retention

Semaphore's per-job artifacts (./report/) are GC'd after roughly 30
days, which doesn't cover the gap between a release tag and the
matching upstream conformance submission for the next k-sigs/gateway-api
release window. Add a third epilogue step that copies the report into
the existing Calico GCS bucket at:

  gs://${GCS_BUILD_CACHE_BUCKET}/gateway-conformance/${ref}/${sha}/

`${ref}` is the tag (release builds) or branch; `${sha}` is the
short commit. PR runs are skipped — the Semaphore artifact is enough
for those, and PR runs would otherwise dominate the bucket.

Bucket-side lifecycle policy is intentionally out of scope. If we want
to expire reports older than N months, that's a one-time gcloud
storage buckets update lifecycle change, not a CI concern. Reusing the
existing transient-build-artifacts bucket inherits its current policy
which is fine for our needs (six-month-ish horizon).

Regenerated semaphore.yml + semaphore-scheduled-builds.yml from the
template.

nelljerram commented on third_party/envoy-gateway/Makefile:
```diff
@@ -2,7 +2,7 @@ include ../../metadata.mk
 PACKAGE_NAME ?= github.com/projectcalico/calico/third_party/envoy-gateway
 
-ENVOY_GATEWAY_IMAGE ?= envoy-gateway
+ENVOY_GATEWAY_IMAGE ?= calico/envoy-gateway
```
It looks like this is against our convention, i.e. other places don't have "calico/" prefix.  Is there a reason for adding it here?

Same for other Envoy images.

electricjesus replied:
The `calico/` prefix is what KIND_CALICO_IMAGES expects (kind-build's loader greps for it to decide what's a Calico-built image vs an upstream pull). I added it on the three Envoy targets so the conformance block can reuse the existing kind-build path without duplicating image-loading logic. Happy to revert and inline a separate loader instead if there's a reason the convention is intentional — I couldn't find one in the surrounding history but I may be missing something. Which way would you prefer?

nelljerram commented on lib.Makefile:
```diff
@@ -1611,6 +1617,18 @@ $(REPO_ROOT)/whisker/.image.created-$(ARCH):
+	$(MAKE) -C $(REPO_ROOT)/third_party/envoy-proxy image
+
+$(REPO_ROOT)/third_party/envoy-ratelimit/.envoy-ratelimit.created-$(ARCH):
+	$(MAKE) -C $(REPO_ROOT)/third_party/envoy-ratelimit image
```
Hmm; probably not an issue for this PR specifically, but looks like we're starting to develop a parallel release image process here, which doesn't feel right.

electricjesus replied:
Yeah I see why it reads that way. To clarify: this is purely the kind-cluster image-load path (KIND_IMAGE_MARKERS files that kind-build uses to decide whether an image needs to be reloaded). It doesn't touch release packaging, tagging, or the published images. The `third_party/envoy-*` directories already produce these images for their own per-PR CI; I'm only adding markers so kind-build knows about them. Open to factoring it out if it bothers you, but I think it's the same shape as the existing KIND_CALICO_IMAGES entries.

caseydavenport replied:
Yeah, this isn't really the same as the release-image flow - this is a dev flow builit specifically for fast iteration against kind (and other clusters too!)

@electricjesus if you rebase on master I think you'll find that the logic for these stamps has changed a little bit to detect when e.g., the stamp exists but the image has been cleaned up under its feet.

nelljerram commented on Makefile:
```diff
@@ -275,6 +275,140 @@ e2e-run-cnp:
+# Caller must:
+#   1. set KUBECONFIG (e.g. $(KIND_KUBECONFIG))
+#   2. set GATEWAY_CONFORMANCE_VERSION to a Calico release tag (e.g. v3.31.5)
+#      for upstream submissions; required by the gateway-api report schema.
```
This detail feels wrong.  Surely this Makefile should be able to infer the current tag?

Shouldn't the overall logic be:
1. Always run the conformance test, as part of our CI.
2. If it passes, AND if we're on a release tag, make an upstream submission to report that.

nelljerram commented on Makefile:
```diff
@@ -275,6 +275,140 @@ e2e-run-cnp:
+GATEWAY_CONFORMANCE_ORG ?= projectcalico
+GATEWAY_CONFORMANCE_PROJECT ?= calico
+GATEWAY_CONFORMANCE_URL ?= https://github.com/projectcalico/calico
+GATEWAY_CONFORMANCE_CONTACT ?= https://github.com/projectcalico/calico/blob/master/GOVERNANCE.md
```
`GOVERNANCE.md` file does not exist.  What is the intent here?

nelljerram commented on e2e/cmd/gateway/manifests/gatewayapi.yaml:
```diff
@@ -0,0 +1,29 @@
+# this by applying this manifest, waiting for the EnvoyProxy CRD,
+# pre-creating the `tigera-gateway` namespace itself, and then
+# applying envoyproxy.yaml so the next reconcile finds the EnvoyProxy
+# and proceeds.
```
A 3rd party EnvoyProxy (for `envoyProxyRef`) should not be configured in the tigera-gateway namespace; and it sound like it would avoid most of the above issue if it wasn't.  Can you tweak to configure the EnvoyProxy in a different namespace instead?

nelljerram commented on e2e/cmd/gateway/manifests/envoyproxy.yaml:
```diff
@@ -0,0 +1,34 @@
+# any envoy pod via IPVS. The Calico tigera-operator picks this up
+# through the GatewayAPI CR's envoyProxyRef (see gatewayapi.yaml).
+#
+# This resource is created in the `tigera-gateway` namespace.
```
As commented elsewhere, I think it should be provisioned in a different namespace.

nelljerram commented on e2e/cmd/gateway/manifests/metallb-pool.yaml:
```diff
@@ -0,0 +1,49 @@
+# get programmed as routes inside the cluster via Calico's BGP fan-out, so
+# in-cluster pods can reach them, but the Gateway API conformance test
+# binary runs from the host (Semaphore VM or developer laptop) where there
+# is no BGP session and the IPs don't resolve.
```
```suggestion
# is no BGP session and so the IPs aren't reachable.
```

nelljerram commented on e2e/cmd/gateway/manifests/metallb-pool.yaml:
```diff
@@ -0,0 +1,49 @@
+# three the upstream conformance suite's base manifests create
+# (gateway-conformance-{infra,app-backend,web-backend}) plus
+# tigera-gateway in case the operator places EG-managed LB Services in
+# the controller's namespace rather than the Gateway's.
```
I believe our Gateway impl always provisions Gateway Services in the tigera-gateway namespace, no?  Because that's what EG does.  I've heard of EG being enhanced to follow the namespace of the Gateway resource, but has that landed yet?

electricjesus replied:
Checked here and you're right `internal/infrastructure/kubernetes/proxy_infra.go::GetResourceNamespace` returns the EG `ControllerNamespace` (so `tigera-gateway` for us) unless `GatewayNamespaceMode` is set. The "follow the namespace of the Gateway resource" enhancement landed (it's `KubernetesDeployModeTypeGatewayNamespace` in v1alpha1, exposed via `Provider.Kubernetes.Deploy.Type`), but it's opt-in. Our tigera-operator's gateway-api render doesn't set it (`Provider.Kubernetes.Deploy` is left at its zero value), so we're firmly in default mode and Services always end up in `tigera-gateway`.

Trimmed the `serviceAllocation.namespaces` list down to just `tigera-gateway` and dropped the hedging comment in 5c98d08a. Verified the conformance run is still green with the trimmed pool.

nelljerram commented on Makefile:
```diff
@@ -275,6 +275,140 @@ e2e-run-cnp:
+# Name of the docker network kind binds to. The kind default is "kind".
+GATEWAY_KIND_DOCKER_NETWORK ?= kind
+# Calico's tigera-operator-provisioned GatewayClass doesn't populate
+# .status.supportedFeatures, so the conformance suite's auto-inference returns
```
Is that a bug that we should just fix in operator, instead of working around?

electricjesus replied:
Resolved a different way after looking at it again. Moved the EnvoyProxy CR out of `tigera-gateway` (which the operator owns) into `e2e-gateway-conformance` (which we own). The deadlock was that the operator's gatewayapi controller resolves every `envoyProxyRef` before it renders the `tigera-gateway` namespace, so referencing it on a fresh install was a chicken-and-egg. With the EnvoyProxy in our own namespace, no workaround is needed. Pushed the change in 095040d9 / 385a7a60 — the namespace pre-create code and the explanatory comments about operator internals are gone.

The underlying operator behaviour is still arguably wrong (a missing EnvoyProxyRef could just requeue instead of going Degraded and returning early), but that's a separate operator-side fix; not in scope here.

nelljerram replied:
What I meant was: should we fix operator so that our GatewayClasses have status.supportedFeatures populated?  It sounds like we are talking at cross-purposes here.

nelljerram commented on Makefile:
```diff
@@ -275,6 +275,140 @@ e2e-run-cnp:
+
+.PHONY: e2e-gateway-setup
+e2e-gateway-setup:
+	@if [ -z "$(KUBECONFIG)" ]; then echo "e2e-gateway-setup: KUBECONFIG must be set"; exit 1; fi
```
I would prefer to delete all the leading `@`s here (and throughout).  IMO more output is helpful.

nelljerram replied:
Also this is a lot of shell to have directly in Makefile.  Worth moving to a script somewhere?

nelljerram commented on Makefile:
```diff
@@ -275,6 +275,140 @@ e2e-run-cnp:
+	@# with "EnvoyProxy ... not found", so the namespace never lands. We unblock
+	@# this by pre-creating the namespace ourselves once the EnvoyProxy CRD is
+	@# in, then dropping the EnvoyProxy CR in. The next reconcile finds it and
+	@# the operator owns the namespace from then on.
```
If the EnvoyProxy is provisioned in another namespace, we still need the CRD.  So I think the sequence becomes:
1. provision GatewayAPI CR
2. wait for CRDs
3. provision EnvoyProxy
4. wait for GatewayClass

so pretty much the same except don't need to pre-provision tigera-gateway namespace.

nelljerram commented on Makefile:
```diff
@@ -275,6 +275,140 @@ e2e-run-cnp:
+	@# in, then dropping the EnvoyProxy CR in. The next reconcile finds it and
+	@# the operator owns the namespace from then on.
+	@end=$$(( $$(date +%s) + $(GATEWAY_SETUP_CRD_TIMEOUT) )); \
+	until kubectl --kubeconfig=$(KUBECONFIG) get crd gatewayclasses.gateway.networking.k8s.io >/dev/null 2>&1 \
```
Don't need to wait for GatewayClass CRD here.

nelljerram commented on Makefile:
```diff
@@ -275,6 +275,140 @@ e2e-run-cnp:
+	  sleep 5; \
+	done
+	@echo "Pre-creating tigera-gateway namespace so the EnvoyProxy CR can be applied before the operator reconciles past the EnvoyProxyRef check"
+	kubectl --kubeconfig=$(KUBECONFIG) create namespace tigera-gateway --dry-run=client -o yaml | kubectl --kubeconfig=$(KUBECONFIG) apply -f -
```
Delete 2 previous lines

nelljerram commented on Makefile:
```diff
@@ -275,6 +275,140 @@ e2e-run-cnp:
+	kubectl --kubeconfig=$(KUBECONFIG) wait --for=condition=Accepted=true --timeout=$(GATEWAY_SETUP_GWC_TIMEOUT) gatewayclass/$(GATEWAY_CLASS_NAME)
+
+## Run the Gateway API conformance suite. Requires e2e-gateway-setup to have completed.
+e2e-run-gateway-conformance:
```
Depend on e2e-gateway-setup?

nelljerram commented on .semaphore/semaphore.yml.d/blocks/20-gateway-conformance.yml:
```diff
@@ -0,0 +1,42 @@
+          - make e2e-gateway-setup KUBECONFIG=$(pwd)/hack/test/kind/kind-kubeconfig.yaml
+          # Run the upstream Gateway API conformance suite against the cluster
+          # and emit the ConformanceReport YAML to ./report/.
+          - .semaphore/run-and-monitor gateway-conformance.log make e2e-run-gateway-conformance KUBECONFIG=$(pwd)/hack/test/kind/kind-kubeconfig.yaml GATEWAY_CONFORMANCE_VERSION=${SEMAPHORE_GIT_TAG_NAME:-${SEMAPHORE_GIT_BRANCH}}
```
Can we replace all the above with `make e2e-test-gateway-conformance` ?  It feels wrong to duplicate the internal steps both here and in Makefile.

nelljerram reviewed (changes requested):
A few detailed comments here, but overall this is looking great.

I'd appreciate another set of eyes on aspects I don't feel confident about:
1. metallb update - does that have effects elsewhere across our Calico product?
2. Adding gateway images to KIND_CALICO_IMAGES - does that have effects elsewhere across our Calico product?

nelljerram commented:
Looking at the gateway-conformance-report.yaml artifact in CI:

> `  organization: projectcalico`
Should it be Tigera?  Or "Project Calico"?  Do you know how this field gets consumed and presented on the Gateway API website?

> `  project: calico`
Ditto; perhaps should be Calico, to look better on the website.

nelljerram commented:
One more nit: I noticed this traceback in the full log:
```
[controller-runtime] log.SetLogger(...) was never called; logs will not be displayed.
Detected at:
	>  goroutine 880 [running]:
	>  runtime/debug.Stack()
	>  	/usr/local/go/src/runtime/debug/stack.go:26 +0x5e
	>  sigs.k8s.io/controller-runtime/pkg/log.eventuallyFulfillRoot()
	>  	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.23.3/pkg/log/log.go:60 +0xcd
	>  sigs.k8s.io/controller-runtime/pkg/log.(*delegatingLogSink).WithValues(0x1730f2c02540, {0x0, 0x0, 0x0})
	>  	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.23.3/pkg/log/deleg.go:168 +0x49
...
```
Would it be easy to clean that up by adding a call to `log.SetLogger`?

electricjesus commented:
thanks @nelljerram ! pinging @caseydavenport for a re-review

electricjesus committed:
refactor(e2e/gateway): move EnvoyProxy CR out of tigera-gateway namespace

Per review feedback: putting the test EnvoyProxy in `tigera-gateway`
forced a chicken-and-egg with the gatewayapi controller's reconcile
order (controller resolves envoyProxyRef before it renders the
operator-owned namespace, so a fresh install hits "EnvoyProxy ... not
found" and never recovers). Moving the EnvoyProxy CR to a test-owned
namespace (`e2e-gateway-conformance`) sidesteps the problem entirely
because we control when that namespace exists.

Manifest comments updated to describe the new bootstrap story instead
of the workaround.

electricjesus committed:
refactor(e2e-gateway-setup): extract setup shell into a script, simplify Makefile

Per review feedback, the e2e-gateway-setup recipe was a wall of inline
shell that's hard to read or run outside Make. Move the body into
hack/test/kind/gateway-setup.sh; the recipe now just exports the
relevant variables and exec's the script. Same observable behaviour,
much easier to debug locally (run the script directly with
KUBECTL=/path/to/kubectl), and the script picks up the kubectl that
ships next to it under hack/test/kind/ when there isn't one in PATH.

While here, fold in a few smaller review items so we don't churn the
target twice:
- e2e-run-gateway-conformance now depends on e2e-gateway-setup directly,
  so e2e-test-gateway-conformance no longer has to call setup itself.
- GATEWAY_CONFORMANCE_VERSION defaults to `git describe --tags --always
  --dirty` instead of being empty by default; the previous "you must
  set this for upstream submission" warning conflated test-runs with
  upstream submission. Whether to submit is a separate decision.
- Implementation identifier on the report changes from `projectcalico`/
  `calico` to `Project Calico`/`Calico` to match how the project styles
  itself on docs.tigera.io and the gateway-api report site.
- Contact URL points at CODE-OF-CONDUCT.md (which exists) instead of
  the never-created GOVERNANCE.md.
- Recipe args are single-quoted so multi-word values like `Project
  Calico` make it through the shell as one argument.

electricjesus committed:
fix(e2e/gateway): silence controller-runtime SetLogger warning

The conformance log was peppered with a noisy stack-trace warning on
every run because controller-runtime emits one the first time any of
its loggers is consulted without SetLogger being called. The upstream
conformance.DefaultOptions installs a zap logger up front for the
same reason; do the same in our runner.

electricjesus committed:
docs(metallb-pool): tighten BGP-reachability comment

Per review suggestion: "the IPs don't resolve" was misleading (the
issue isn't DNS, it's L3 reachability). Reword to "the IPs aren't
reachable", which is what we actually mean.

electricjesus committed:
ci(gateway-conformance): collapse block to invoke make e2e-test-gateway-conformance

Per review feedback the block was duplicating the four-step recipe
(build runner, kind-up, e2e-gateway-setup, e2e-run-gateway-conformance)
that already exists as e2e-test-gateway-conformance in the Makefile.
Replace the inline commands with a single make invocation so the
locally-runnable target and the CI block stay in lockstep automatically.

Also extends change_in() to include hack/test/kind/gateway-setup.sh
(the new home for the setup shell) so edits there re-trigger the
block.

Regenerated semaphore.yml and semaphore-scheduled-builds.yml from the
template.

electricjesus commented:
Thanks for the careful read Nell. On the two points you flagged not feeling confident about — both are scoped strictly to the kind CI flow, nothing changes for shipped Calico:

The metallb upgrade only touches files under `hack/test/kind/`. Production Calico doesn't bundle metallb at all (it's purely a kind-cluster convenience for assigning LoadBalancer IPs in CI). I had to upgrade because metallb 0.14 dropped the legacy ConfigMap install path — the new metallb-config.yaml is a straight 1:1 conversion of the old ConfigMap (same `80.15.0.0/24` + `fdff::/64` ranges, same BGPAdvertisement). I checked deploy_resources.sh and the kind infra layout to make sure nothing else was leaning on the ConfigMap shape. So no spillover.

The KIND_CALICO_IMAGES extension is similarly local. The three `calico/envoy-{gateway,proxy,ratelimit}` images already get built per-PR by the existing Envoy Gateway / Envoy Proxy / Envoy Ratelimit blocks; this change just teaches kind-build to load them so the conformance block has them on the kind nodes. They aren't released, packaged, or tagged differently. Happy to walk through the call graph if helpful.

The reason both pieces of plumbing existed at all is that Calico's kind topology diverges from Envoy Gateway's own conformance CI, which is also why a chunk of the per-test workarounds were needed. Posting a side-by-side because I think it'll help future-me too:

Calico kind: 1 control + 3 workers. Envoy Gateway kind: `NUM_WORKERS=0`, single node. That single difference is why we needed the L2 metallb pool from the kind subnet (single-node EG kind never has cross-node speaker/backend mismatch) and why we override `externalTrafficPolicy` to `Cluster` on the Envoy data-plane Service (`Local` works fine for them because speaker = backend node always; on multi-node we drop packets in IPVS).

Calico install path: tigera-operator GatewayAPI CR. Envoy Gateway CI: their own Helm chart, direct. That's why the bootstrap order is more involved on our side. After your review I moved the EnvoyProxy CR out of `tigera-gateway` and into a test-owned namespace — the operator's gatewayapi controller renders namespaces after resolving every `envoyProxyRef`, so referencing `tigera-gateway` on a fresh install was a chicken-and-egg. With the EnvoyProxy in our own namespace, no pre-create dance is needed and the comments calling out an operator quirk are gone. (Still worth filing a separate operator issue: the controller could just requeue on missing EnvoyProxy instead of going Degraded, but that's not in scope for this PR.)

Outside those two roots, everything else is a thin wrapper around the same upstream `sigs.k8s.io/gateway-api/conformance/utils/suite` that EG itself uses. The curated feature/test/profile set is mirrored verbatim from envoyproxy/gateway v1.7.0's `internal/gatewayapi/conformance/suite.go` — so what we claim is what they claim, minus what's not relevant for us (mesh).

electricjesus committed:
docs(e2e/gateway): align report org/project with submission convention

Surveyed existing submissions in
kubernetes-sigs/gateway-api/conformance/reports/v1.3.0/. Established
implementations all use lowercase, hyphen-separated identifiers in
the YAML's `organization` and `project` fields:

  envoyproxy / envoy-gateway
  cilium     / cilium
  aws        / aws-load-balancer-controller
  GKE        / gke-gateway        (sole TitleCase outlier)

These values become the directory keys under conformance/reports/.
The implementations index on gateway-api.sigs.k8s.io is hand-curated,
so display niceness doesn't depend on the YAML.

Calico's original value was already in line with the convention and
the GitHub org name (github.com/projectcalico/calico). Revert to
projectcalico / calico.

electricjesus commented:
@nelljerram Re: logger, fixed in 41ae1a91 — the runner now installs a zap logger up front, mirroring upstream `conformance.DefaultOptions`. The traceback is gone from subsequent runs.

electricjesus commented:
@nelljerram good catch on this one, the answer surprised me. I went and looked at the existing submissions in `kubernetes-sigs/gateway-api/conformance/reports/v1.3.0/` and the convention is actually lowercase, hyphenated. Examples:

- envoyproxy / envoy-gateway
- cilium / cilium
- aws / aws-load-balancer-controller
- GKE / gke-gateway (sole TitleCase outlier)

These values become the directory key under `conformance/reports/<gw-api-version>/<org>-<project>/`. The implementations index on gateway-api.sigs.k8s.io is hand-curated, so the display niceness doesn't depend on the YAML.

So the original `projectcalico` / `calico` was already in line — matches the convention, matches the GitHub org name. Reverted in 5081bbb6.

electricjesus committed:
docs(metallb-pool): scope serviceAllocation to tigera-gateway only

Previously the pool's serviceAllocation listed all four namespaces a
conformance run touches (gateway-conformance-{infra,app-backend,
web-backend} plus tigera-gateway), with a comment hedging "in case the
operator places EG-managed LB Services in the controller's namespace
rather than the Gateway's." That hedge was unnecessary.

Verified against EG source (internal/infrastructure/kubernetes/
proxy_infra.go GetResourceNamespace): when GatewayNamespaceMode is
unset, EG provisions data-plane Services in the EG ControllerNamespace.
Verified against the tigera-operator gateway-api render: it leaves
Provider.Kubernetes.Deploy unset, so GatewayNamespaceMode is false. So
LB Services always land in tigera-gateway for our setup.

The Gateway CRs themselves are in gateway-conformance-infra, but those
are CRs, not LB Services. metallb only cares about the latter.

Verified on the VM that trimming the namespace list to just
tigera-gateway still produces a clean green conformance run.

caseydavenport commented:
Just noting that I am planning to get to this one, but it might be a bit later in the week! Thanks @electricjesus !

nelljerram commented on e2e/cmd/gateway/manifests/envoyproxy.yaml:
```diff
@@ -0,0 +1,32 @@
+    type: Kubernetes
+    kubernetes:
+      envoyService:
+        externalTrafficPolicy: Cluster
```
This is fine, but FWIW we now also have the option of using the new Patch field.  (Although that would make this harder to backport.)

nelljerram reviewed (approved):
I'm happy with this now.  Couple more comments below, but non-blocking.

electricjesus commented:
thanks @nelljerram ! 


RE:

> What I meant was: should we fix operator so that our GatewayClasses have status.supportedFeatures populated? It sounds like we are talking at cross-purposes here.

looks like the empty `status.supportedFeatures` actually isn't tigera-operator's omission but envoyproxy/gateway deliberately leaves it unpopulated:
- 1.7.2 https://github.com/envoyproxy/gateway/blob/release/v1.7/internal/gatewayapi/status/gatewayclass.go#L36-L41
- also in main https://github.com/envoyproxy/gateway/blob/main/internal/gatewayapi/status/gatewayclass.go#L36 

And they cite the field's experimental/unstable datatype as the reason

So the `-supported-features` override in the Makefile is the same workaround upstream's own conformance CI uses, not Calico-specific.

I'll update the comment in the Makefile to make that clearer. If/when upstream re-enables population once the datatype graduates, we'll inherit it.

electricjesus committed:
docs(makefile): clarify upstream EG omits supportedFeatures by design

nelljerram commented:
> thanks @nelljerram !
> 
> RE:
> 
> > What I meant was: should we fix operator so that our GatewayClasses have status.supportedFeatures populated? It sounds like we are talking at cross-purposes here.
> 
> looks like the empty `status.supportedFeatures` actually isn't tigera-operator's omission but envoyproxy/gateway deliberately leaves it unpopulated:
> 
> * 1.7.2 https://github.com/envoyproxy/gateway/blob/release/v1.7/internal/gatewayapi/status/gatewayclass.go#L36-L41
> * also in main https://github.com/envoyproxy/gateway/blob/main/internal/gatewayapi/status/gatewayclass.go#L36
> 
> And they cite the field's experimental/unstable datatype as the reason
> 
> So the `-supported-features` override in the Makefile is the same workaround upstream's own conformance CI uses, not a Calico-specific.
> 
> I'll update the comment in the Makefile to make that clearer. If/when upstream re-enables population once the datatype graduates, we'll inherit it.

Ah right.  I was imagining that tigera/operator could populate status.supportedFeatures at the time that it provisions the GatewayClass - but I guess not, because we should only populate `spec`, and it makes sense that upstream is responsible for `status`

electricjesus committed:
fix(third_party): drop redundant calico/ prefix on envoy BUILD_IMAGE

The retag-build-images-with-registries step (called after each envoy
sub-make's docker build) already prefixes REGISTRY=calico, producing
calico/envoy-{gateway,proxy,ratelimit}:latest-$ARCH for KIND_CALICO_IMAGES
to pick up. Carrying the prefix on BUILD_IMAGE itself was redundant and
also leaked into release/internal/pinnedversion's image catalogue, breaking
its approval-snapshot tests (TestImageComponents, TestGeneratePinnedVersionFile).

caseydavenport commented on e2e/cmd/gateway/manifests/envoyproxy.yaml:
```diff
@@ -0,0 +1,32 @@
+# conformance CI gets away with this because it uses a single-node kind
+# cluster (tools/hack/create-cluster.sh defaults NUM_WORKERS to 0), so
+# the metallb speaker that wins the IP and the envoy backend pod always
+# share a node. Calico's kind setup is multi-node (1 control + 3
```
We do support a single-node kind setup that might be helpful here? libcalico-go spins it up - there's a kind config env var that controls it. 

Multi-node is fine too, just a bit more resource / time intensive.

caseydavenport commented on hack/test/kind/deploy_resources.sh:
```diff
@@ -187,9 +187,27 @@ ${kubectl} apply -f ${INFRA_DIR}/calicoctl.yaml
 ${kubectl} create ns metallb-system || true
 ${kubectl} apply -f ${INFRA_DIR}/metallb.yaml
-${kubectl} apply -f ${INFRA_DIR}/metallb-config.yaml
+# metallb-native installs CRDs and a ValidatingWebhookConfiguration; wait
```
I do wonder if we should just be moving away from metal LB altogether in this test suite - IIUC we have our own LoadBalancer IP allocation implementation now that is available without needing to install any extra addons.

Probably a cleaner approach?

electricjesus replied:
Done — Calico's kube-controllers loadbalancer controller is now the IPAM authority (Calico-native IPPools in `hack/test/kind/infra/calico-lb-pools.yaml`, `allowedUses: [LoadBalancer]`). metallb is still installed, but only to serve the Gateway API conformance L2 pool — Calico has no L2 announcer (confd is BGP-only, the LB controller is IPAM-only), so the conformance host runner still needs metallb's L2 speaker to ARP for LB IPs on the kind docker bridge. Validated on a build VM: node k8st BGP 24/24, Gateway API conformance 430/430.

caseydavenport commented on node/tests/k8st/tests/test_bgp_advert.py:
```diff
@@ -764,8 +764,8 @@ def test_bgp_filter_ip_advertisement(self):
 """)
-        kubectl("patch --type=merge bgppeer node-extra.peer --patch '{\"spec\": {\"filters\": [\"test-filter-export-1\"]}}'")
-        self.add_cleanup(lambda: kubectl("patch --type=merge bgppeer node-extra.peer --patch '{\"spec\": {\"filters\": []}}'"))
+        kubectl("patch --type=merge bgppeer.crd.projectcalico.org node-extra.peer --patch '{\"spec\": {\"filters\": [\"test-filter-export-1\"]}}'")
```
This is wrong - we shouldn't be referring to this API group here. Why did you need this change?

electricjesus replied:
Fixed — pushed `.projectcalico.org` (v3 group) everywhere. Qualification is still needed because metallb v0.14 ships `bgppeers.metallb.io`, so `bgppeer` shortname resolves ambiguously now. Stripping that CRD from the manifest crashes metallb's controller at startup (it pre-resolves every type at boot).

Same push tackles your "we have our own LB allocator now" point: 80.15.0.0/24 and fdff::/64 are now Calico-native IPPools (`allowedUses: [LoadBalancer]`); IPv6 single-IP LB test converges on the IPv4 sibling's pattern. metallb stays installed only for Gateway API conformance L2 reachability — confd is BGP-only and kube-controllers' LB controller is IPAM-only, so metallb's L2 speaker remains the only LB-IP L2 announcer in our stack.

Validated on a build VM: node k8st BGP 24/24, Gateway API conformance 430/430.

caseydavenport replied:
Must we install the bgppeers CRD from metallb?

caseydavenport replied:
Yeah, seems like we must.

caseydavenport commented on lib.Makefile:
```diff
@@ -1711,6 +1717,26 @@ $(REPO_ROOT)/key-cert-provisioner/.image.created-$(ARCH): \
+# actual built images, so it can run in parallel with component builds.
+# Track the operator source selection in a generated inputs file so
+# changes to repo/branch invalidate the operator stamp.
+.PHONY: FORCE
```
What's going on here? Looks like dead code?

electricjesus replied:
done removed

caseydavenport reviewed (commented):
A few thoughts / questions.

electricjesus committed:
fix(lib.Makefile): drop dead FORCE target

Introduced in 87ef9fc88d but never referenced anywhere as a
prerequisite. The accompanying comment block referred to the
operator-build stamp logic that was reworked in the meantime, so it
also no longer fit. Remove both.

electricjesus committed:
test(kind): move default LB IP allocation off metallb to Calico-native

The metallb v0.14 upgrade left two pain points:

1. metallb's BGPPeer CRD (bgppeers.metallb.io) collides with Calico's
   BGPPeer in kubectl shortname resolution, breaking node BGP tests'
   `kubectl ... bgppeer node-extra.peer` commands.

2. metallb is doing IPAM for IPs that Calico already advertises via BGP
   (see test_bgp_advert.py:561 -- the BGPConfiguration.serviceLoadBalancerIPs
   allow-list already covers 80.15.0.0/24). The IPv4 single-IP LB test
   at test_bgp_advert.py:936-947 already uses pure Calico-native LB IPAM
   (IPPool with allowedUses: [LoadBalancer] + projectcalico.org/loadBalancerIPs
   annotation); the IPv6 sibling at test_bgp_advert_v6.py:695 is the only
   remaining test that still uses metallb's annotation.

Move the cluster-default LB IPAM authority (80.15.0.0/24 + fdff::/64) to
kube-controllers' loadbalancer controller, delete metallb-config.yaml,
and converge the IPv6 single-IP LB test on the IPv4 sibling's pattern.
Qualify the kubectl `bgppeer` shortname with `.projectcalico.org` (the
v3 user-facing API group, not the CRD storage group) in the affected
node BGP tests to avoid the metallb collision.

LB Services in node k8st tests now declare `spec.loadBalancerClass: calico`
explicitly. Without that, metallb's ServiceReconciler (still installed
for Gateway API L2 reachability) treats unclassified LB Services as its
own, races Calico's status updates, and overwrites them with empty
status. Pinning the class to "calico" makes metallb back off
(`IsCalicoManagedLoadBalancer` matches; metallb's class-match doesn't).

Why metallb stays installed: it's still the only L2 announcer in our
stack -- the kube-controllers loadbalancer controller is IPAM-only (no
ARP/L2 code) and confd advertises only via BGP. Gateway API conformance
runs from the host and needs an LB IP that's L2-reachable on the kind
docker bridge, so metallb's L2 speaker is currently irreplaceable.
The metallb manifest is kept verbatim because the controller refuses to
start when any of its CRDs are missing -- it builds a controller-runtime
manager that pre-resolves the namespacedness of every type at startup
(see metallb v0.14.9 internal/k8s/k8s.go:160). A partial-strip variant
crashed with `failed to determine if *v1beta1.Community is namespaced`.

Changes:
- New hack/test/kind/infra/calico-lb-pools.yaml: two IPPools with
  allowedUses: [LoadBalancer] applied after Calico reports ready
- Delete hack/test/kind/infra/metallb-config.yaml (default BGP pool +
  BGPAdvertisement); deploy_resources.sh applies the Calico pools
  after tigerastatus is Available, and waits only on the L2-mode CRDs
- node/tests/k8st/test_base.py:create_service: emit
  spec.loadBalancerClass: calico for type=LoadBalancer to disambiguate
  ownership between Calico's LB controller and metallb's reconciler
- node/tests/k8st/tests/test_bgp_advert{,_v6}.py: same loadBalancerClass
  on the inline-YAML LB Services; swap the IPv6 test's metallb
  annotation for projectcalico.org/loadBalancerIPs to mirror the IPv4
  sibling at test_bgp_advert.py:957
- node/tests/k8st/tests/test_bgp_advert.py: drop the inline 80.15.0.100/32
  IPPool from test_single_ip_lb_rr; the new cluster-default lb-pool-v4
  covers this range and IPPool CIDRs aren't allowed to overlap (see
  libcalico-go/lib/clientv3/ippool.go:498)
- node/tests/k8st/tests/test_bgp_{filter,advert,advert_v6}.py: qualify
  `kubectl ... bgppeer` with `.projectcalico.org` to disambiguate from
  metallb's same-shortname CRD
- e2e/cmd/gateway/manifests/metallb-pool.yaml: header rewritten to
  document why the L2 pool stays metallb-based
